File: ruby/test/test_queue.rb
require 'helper'
require 'mongo'
require 'karait'

class TestQueue < Test::Unit::TestCase
  
  def setup
    Mongo::Connection.new()['karait_test']['queue_test'].drop()
  end
  
  should "initialize a capped collection when a queue is created" do
    queue = Karait::Queue.new(
      :database => 'karait_test',
      :queue => 'queue_test',
      :average_message_size => 8192,
      :queue_size => 4096
    )
    
    options = Mongo::Connection.new()['karait_test']['queue_test'].options
    
    assert_equal true, options['capped']
    assert_equal 4096, options['max']
    assert_equal (8192 * 4096), options['size']
  end
  
  should "attach to a mongo queue collection that already exists" do
    collection = Mongo::Connection.new()['karait_test']['queue_test']
    collection.insert({
      :message => {
        :apple => 3,
        :banana => 5
      },
      :_meta => {
        :timestamp => 2523939,
        :expire => 20393,
        :routing_key => 'foo_key'
      }
    })
    assert_equal 1, collection.count()
    
    queue = Karait::Queue.new(
      :database => 'karait_test',
      :queue => 'queue_test'
    )

    collection = Mongo::Connection.new()['karait_test']['queue_test']
    assert_equal 1, collection.count()
    assert_equal nil, collection.options()
  end
  
  should "write a dictionary into the mongo queue collection" do
    queue = Karait::Queue.new(
      :database => 'karait_test',
      :queue => 'queue_test',
      :average_message_size => 8192,
      :queue_size => 4096
    )
    
    queue.write({
      :apple => 5,
      :banana => 6,
      :inner_object => {
        :foo => 1,
        :bar => 2
      }
    })

    collection = Mongo::Connection.new()['karait_test']['queue_test']
    obj = collection.find_one()
    assert_equal 6, obj['banana']
    assert_equal 2, obj['inner_object']['bar']
    assert obj['_meta']['expire']
    assert obj['_meta']['timestamp']
  end
  
  should "write a message object into the mongo queue collection" do
    queue = Karait::Queue.new(
      :database => 'karait_test',
      :queue => 'queue_test'
    )
    message = Karait::Message.new
    message.apple = 5
    message.banana = {
      :foo => 2
    }
    queue.write message
    
    collection = Mongo::Connection.new()['karait_test']['queue_test']
    obj = collection.find_one
    assert_equal 5, obj['apple']
    assert_equal 2, obj['banana']['foo']
  end

  should "read a message object from the mongo queue collection" do
    queue = Karait::Queue.new(
      :database => 'karait_test',
      :queue => 'queue_test'
    )
  
    write_message = Karait::Message.new
    write_message.apple = 5
    write_message.banana = 6
    write_message.inner_object = {
      'foo' => 1,
      'bar' => 2
    }
    queue.write write_message
  
    read_message = queue.read()[0]
    assert_equal 5, read_message.apple
    assert_equal 2, read_message.inner_object['bar']
    assert_equal 3, read_message.to_hash.keys.count
  end
  
  should "return messages in fifo order" do
    queue = Karait::Queue.new(
      :database => 'karait_test',
      :queue => 'queue_test'
    )
    
    queue.write Karait::Message.new({'foo' => 1})
    queue.write Karait::Message.new({:foo => 2})
    queue.write Karait::Message.new({'foo' => 3})
    messages = queue.read()
    
    assert_equal 1, messages[0].foo
    assert_equal 2, messages[1].foo
    assert_equal 3, messages[2].foo
  end
  
  should "only return messages with the appropriate routing key when it's provided" do
    queue = Karait::Queue.new(
      :database => 'karait_test',
      :queue => 'queue_test'
    )
    
    queue.write(Karait::Message.new({:foo => 1}), :routing_key => 'foobar')
    queue.write Karait::Message.new({:foo => 2})

    messages = queue.read(:routing_key => 'foobar')
    assert_equal 1, messages.count
    assert_equal 1, messages[0].foo
  end
  
  should "only return messages with no routing key when none is provided" do
    queue = Karait::Queue.new(
      :database => 'karait_test',
      :queue => 'queue_test'
    )
    
    queue.write(Karait::Message.new({:foo => 1}), :routing_key => 'foobar')
    queue.write Karait::Message.new({:foo => 2})
    
    messages = queue.read()
    assert_equal 1, messages.count
    assert_equal 2, messages[0].foo
  end
  
  should "should no longer return a message when delete is called on it" do
    queue = Karait::Queue.new(
      :database => 'karait_test',
      :queue => 'queue_test'
    )
    
    queue.write Karait::Message.new({:foo => 1})
    messages = queue.read()
    assert_equal 1, messages.count
    assert_equal 1, messages[0].foo
    
    messages[0].delete
    messages = queue.read()
    assert_equal 0, messages.count
  end

  should "not remove message immediately with expire set" do
    queue = Karait::Queue.new(
      :database => 'karait_test',
      :queue => 'queue_test'
    )
    
    queue.write(Karait::Message.new({:foo => 1}), :expire => 0.5)
    sleep(0.1)
    messages = queue.read()
    assert_equal 1, messages.count
  end
  
  should "remove message once expire time is passed" do
    queue = Karait::Queue.new(
      :database => 'karait_test',
      :queue => 'queue_test'
    )
    
    queue.write(Karait::Message.new({:foo => 1}), :expire => 0.1)
    sleep(0.2)
    messages = queue.read()
    assert_equal 0, messages.count
    
    # Make sure the meta._expired key is actually set.
    collection = Mongo::Connection.new()['karait_test']['queue_test']
    raw_message = collection.find_one
    assert_equal true, raw_message['_meta']['expired']
  end
  
  
  should "remove all messages in array when delete messages called" do
    queue = Karait::Queue.new(
      :database => 'karait_test',
      :queue => 'queue_test'
    )
    
    queue.write Karait::Message.new({'foo' => 1})
    queue.write Karait::Message.new({:foo => 2})
    queue.write Karait::Message.new({'foo' => 3})
    messages = queue.read()
    queue.delete_messages messages
    messages = queue.read()
    assert_equal 0, messages.count
  end
  
  should "not see message in queue again until visibility timeout has passed" do
    queue = Karait::Queue.new(
      :database => 'karait_test',
      :queue => 'queue_test'
    )
    
    queue.write Karait::Message.new({'foo' => 1})
    queue.write Karait::Message.new({:foo => 2})
    queue.write Karait::Message.new({'foo' => 3})
    messages = queue.read(:visibility_timeout => 0.4)
    assert_equal 3, messages.count
    messages = queue.read(:visibility_timeout => 0.4)
    assert_equal 0, messages.count
    sleep(0.5)
    messages = queue.read(:visibility_timeout => 0.4)
    assert_equal 3, messages.count
  end
  
  should "block on reading from queue until polling timeout reached" do
    start_time = Time.new.to_f
    
    queue = Karait::Queue.new(
      :database => 'karait_test',
      :queue => 'queue_test'
    )
    
    queue.read( :block => true, :polling_timeout => 0.1 )
    queue.write Karait::Message.new({'foo' => 1})
    messages = queue.read( :block => true, :polling_timeout => 0.1 )

    assert_equal 1, messages.count
    stop_time = Time.new.to_f
    assert_equal true, stop_time - start_time > 0.1
  end
end

File: node/test/test_queue.js
var equal = require('assert').equal,
    puts = require('sys').puts,
    Queue = require('../lib').Queue,
    Message = require('../lib').Message,
    mongodb = require('mongodb'),
    Db = mongodb.Db,
    Connection = mongodb.Connection,
    Server = mongodb.Server;
    
exports.tests = {
    'should initialize a capped collection when a queue is created': function(finished, prefix) {
        new Queue(
            {
                database: 'karait_test',
                queue: 'queue_test',
                averageMessageSize: 8192,
                queueSize: 4096
            },
            function(err, queue) {
                queue.queueCollection.options(function(err, options) {
                    equal(true, options.capped, prefix + 'collection not capped');
                    equal(4096, options.max, prefix + 'invalid max queue size');
                    equal( (4096 * 8192) , options.size, prefix + 'invalid queue size');
                    
                    finished();
                });
            }
        );
    },
    
    'should attach to a collection that already exists': function(finished, prefix) {
        var db = new Db('karait_test', new Server('localhost', 27017, {}), {native_parser:false});
        db.open(function(err, db) {
            db.createCollection('queue_test', {}, function(err, collection) {
                collection.insert(
                    {
                        message: {
                          apple: 3,
                          banana: 5
                        },
                        _meta: {
                          timestamp: 2523939,
                          expire: 20393,
                          routing_key: 'foo_key'
                        }
                    },
                    {
                        safe: true
                    }
                );
                
                new Queue(
                    {
                        database: 'karait_test',
                        queue: 'queue_test',
                        averageMessageSize: 8192,
                        queueSize: 4096
                    },
                    function(err, queue) {
                        queue.queueCollection.options(function(err, options) {
                            collection.count({}, function(err, count) {
                                equal(1, count, prefix + 'count should be 1.');
                                finished();
                            });
                        });
                    }
                );
                
            });
        });
    },

    'should write a dictionary into the mongodb queue collection': function(finished, prefix) {
        new Queue(
            {
                database: 'karait_test',
                queue: 'queue_test',
                averageMessageSize: 8192,
                queueSize: 4096
            },
            function(err, queue) {
                queue.write(
                    {
                        foo: 'bar'
                    }, 
                    function() {
                        var db = new Db('karait_test', new Server('localhost', 27017, {}), {native_parser:false});
                        db.open(function(err, db) {
                            db.createCollection('queue_test', {}, function(err, collection) {
                                collection.findOne({}, function(err, rawMessage) {
                                    equal('bar', rawMessage.foo, prefix + 'foo key not set');
                                    equal(true, rawMessage._meta.timestamp > 0, prefix + 'timestamp was not set');
                                    finished();
                                });
                            });
                        });
                    }
                );
            }
        );
    },
    
    'should write a message object into the mongodb queue collection': function(finished, prefix) {
        new Queue(
            {
                database: 'karait_test',
                queue: 'queue_test',
                averageMessageSize: 8192,
                queueSize: 4096
            },
            function(err, queue) {
                var message = new Message(
                    {
                        'bar': 'foo'
                    }
                );
                
                queue.write(
                    message,
                    function() {
                        var db = new Db('karait_test', new Server('localhost', 27017, {}), {native_parser:false});
                        db.open(function(err, db) {
                            db.createCollection('queue_test', {}, function(err, collection) {
                                collection.findOne({}, function(err, rawMessage) {
                                    var count = 0;
                                    for (var key in rawMessage) {
                                        count += 1;
                                    }
                                    equal('foo', rawMessage.bar, prefix + 'foo key not set');
                                    equal(3, count, prefix + 'rawMessage had wrong number of keys');
                                    equal(true, rawMessage._meta.timestamp > 0, prefix + 'timestamp was not set');
                                    finished();
                                });
                            });
                        });
                    }
                );
            }
        );
    },
    
    'should read a messge object from the mongo queue collection': function(finished, prefix) {
        new Queue(
            {
                database: 'karait_test',
                queue: 'queue_test',
                averageMessageSize: 8192,
                queueSize: 4096
            },
            function(err, queue) {
                writeMessage = new Message({
                   foo: 1,
                   bar: 2,
                   innerObject: {
                       apple: 3
                   }
                });
                queue.write(writeMessage, {}, function() {
                    queue.read(function(err, messages) {
                        var readMessage = messages[0];
                        equal(1, readMessage.foo, prefix + 'foo not set');
                        equal(3, readMessage.innerObject.apple, prefix + 'inner object not found');
                        finished();
                    });
                });
            }
        );
    },
    
    'should only return messages that match routing key': function(finished, prefix) {
        new Queue(
            {
                database: 'karait_test',
                queue: 'queue_test',
                averageMessageSize: 8192,
                queueSize: 4096
            },
            function(err, queue) {
                queue.write({foo: 'bar'}, {routingKey: 'foobar'}, function() {
                    queue.write({bar: 'foo'}, function() {
                        queue.read({routingKey: 'foobar'}, function(err, messages) {
                            equal(1, messages.length, prefix + messages.length + ' not equal to 1');
                            equal('bar', messages[0].foo, prefix + messages[0].foo + ' not equal to bar');
                            queue.read(function(err, messages) {
                                equal(1, messages.length, prefix + messages.length + ' not equal to 1');
                                equal('foo', messages[0].bar, prefix + messages[0].bar + ' not equal to foo');
                                finished();
                            });
                        });
                    });
                });
            }
        );
    },
    
    'should no longer return a message when delete is called on it': function(finished, prefix) {
        new Queue(
            {
                database: 'karait_test',
                queue: 'queue_test',
                averageMessageSize: 8192,
                queueSize: 4096
            },
            function(err, queue) {
                queue.write({foo: 'bar'}, function() {
                    queue.read(function(err, messages) {
                        equal(1, messages.length, prefix + messages.length + ' not equal to 1');
                        messages[0].delete(function() {
                            queue.read(function(err, messages) {
                                equal(0, messages.length, prefix + messages.length + ' not equal to 0');
                                finished();
                            });
                        });
                    });
                });
            }
        );
    },
    
    'should not return expired messages when read is called': function(finished, prefix) {
        new Queue(
            {
                database: 'karait_test',
                queue: 'queue_test',
                averageMessageSize: 8192,
                queueSize: 4096
            },
            function(err, queue) {
                queue.write({foo: 'bar'}, {expire: 0.1}, function() {
                    setTimeout(function() {
                        queue.read(function(err, messages) {
                            equal(0, messages.length, prefix + messages.length + ' not equal to 0');
                            finished();
                        });
                    }, 200);
                });
            }
        );
    },
    
    'should not immediately expire a message when written if expire set': function(finished, prefix) {
        new Queue(
            {
                database: 'karait_test',
                queue: 'queue_test',
                averageMessageSize: 8192,
                queueSize: 4096
            },
            function(err, queue) {
                queue.write({foo: 'bar'}, {expire: 0.1}, function() {
                    setTimeout(function() {
                        queue.read(function(err, messages) {
                            equal(1, messages.length, prefix + messages.length + ' not equal to 1');
                            finished();
                        });
                    }, 1);
                });
            }
        );
    },
    
    'should remove all messages when deleteMessages() is called with an array of messages': function(finished, prefix) {
        new Queue(
            {
                database: 'karait_test',
                queue: 'queue_test',
                averageMessageSize: 8192,
                queueSize: 4096
            },
            function(err, queue) {
                queue.write({foo: 'bar'}, function() {
                    queue.write({foo: 'bar'}, function() {
                        queue.read(function(err, messages) {
                            equal(2, messages.length, prefix + messages.length + ' not equal to 2');
                            queue.deleteMessages(messages, function() {
                                queue.read(function(err, messages) {
                                    equal(0, messages.length, prefix + messages.length + ' not equal to 0');
                                    finished();
                                });
                            });
                        });
                    });
                });
            }
        );
    },
    
    'should not see a message until the visibility timeout has passed': function(finished, prefix) {
        new Queue(
            {
                database: 'karait_test',
                queue: 'queue_test',
                averageMessageSize: 8192,
                queueSize: 4096
            },
            function(err, queue) {
                queue.write({foo: 'bar'}, function() {
                    queue.read({visibilityTimeout: 0.1}, function(err, messages) {
                        equal('bar', messages[0].foo, prefix + messages[0].foo + ' not equal to bar');
                        equal(1, messages.length, prefix + messages.length + ' not equal to 1');
                        queue.read(function(err, messages) {
                            equal(0, messages.length, prefix + messages.length + ' not equal to 0');
                            setTimeout(function() {
                                queue.read(function(err, messages) {
                                    equal(1, messages.length, prefix + messages.length + ' not equal to 1');
                                    finished();
                                });
                            }, 200);
                        });
                    });
                });
            }
        );
    }
};